File: VB/FileSystemData/BusinessObjects/FileSystemLinkObject.vb
Imports System
Imports System.IO
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports System.ComponentModel
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Validation

Namespace FileSystemData.BusinessObjects
	''' <summary>
	''' This class enables you to add soft links to real files instead of saving their contents to the database. It is intended for use in Windows Forms applications only.
	''' </summary>
	<DefaultProperty("FileName")>
	Public Class FileSystemLinkObject
		Inherits BaseObject
		Implements IFileData, IEmptyCheckable, ISupportFullName

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		#Region "IFileData Members"
		<Size(260), Custom("AllowEdit", "False")>
		Public Property FileName() As String Implements IFileData.FileName
			Get
				Return GetPropertyValue(Of String)("FileName")
			End Get
			Set(ByVal value As String)
				SetPropertyValue("FileName", value)
			End Set
		End Property
		Private Sub IFileData_Clear() Implements IFileData.Clear
			Size = 0
			FileName = String.Empty
		End Sub
		'Dennis: Fires when uploading a file.
		Private Sub IFileData_LoadFromStream(ByVal fileName As String, ByVal source As Stream) Implements IFileData.LoadFromStream
			Size = CInt(source.Length)
			Me.FileName = fileName
		End Sub
		'Dennis: Fires when saving or opening a file.
		Private Sub IFileData_SaveToStream(ByVal destination As Stream) Implements IFileData.SaveToStream
			Try
				If destination Is Nothing Then
					FileSystemDataModule.OpenFileWithDefaultProgram(FullName)
				Else
					FileSystemDataModule.CopyFileToStream(FullName, destination)
				End If
			Catch exc As Exception
				Throw New UserFriendlyException(exc)
			End Try
		End Sub
		Private ReadOnly Property IFileData_Size() As Integer Implements IFileData.Size
			Get
				Return Size
			End Get
		End Property
		<Persistent>
		Public Property Size() As Integer
			Get
				Return GetPropertyValue(Of Integer)("Size")
			End Get
			Private Set(ByVal value As Integer)
				SetPropertyValue(Of Integer)("Size", value)
			End Set
		End Property
		#End Region
		#Region "IEmptyCheckable Members"
		Public ReadOnly Property IsEmpty() As Boolean Implements IEmptyCheckable.IsEmpty
			Get
				Return Not File.Exists(FullName)
			End Get
		End Property
		#End Region
		#Region "ISupportFullName Members"
		<Size(260), Custom("AllowEdit", "False")>
		Public Property FullName() As String Implements ISupportFullName.FullName
			Get
				Return GetPropertyValue(Of String)("FullName")
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)("FullName", value)
			End Set
		End Property
		#End Region
	End Class
End Namespace

File: CS/FileSystemData/BusinessObjects/FileSystemLinkObject.cs
using System;
using System.IO;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using System.ComponentModel;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;

namespace FileSystemData.BusinessObjects {
    /// <summary>
    /// This class enables you to add soft links to real files instead of saving their contents to the database. It is intended for use in Windows Forms applications only.
    /// </summary>
    [DefaultProperty("FileName")]
    public class FileSystemLinkObject : BaseObject, IFileData, IEmptyCheckable, ISupportFullName {
        public FileSystemLinkObject(Session session) : base(session) { }
        #region IFileData Members
        [Size(260), Custom("AllowEdit", "False")]
        public string FileName {
            get { return GetPropertyValue<string>("FileName"); }
            set { SetPropertyValue("FileName", value); }
        }
        void IFileData.Clear() {
            Size = 0;
            FileName = string.Empty;
        }
        //Dennis: Fires when uploading a file.
        void IFileData.LoadFromStream(string fileName, Stream source) {
            Size = (int)source.Length;
            FileName = fileName;
        }
        //Dennis: Fires when saving or opening a file.
        void IFileData.SaveToStream(Stream destination) {
            try {
                if (destination == null)
                    FileSystemDataModule.OpenFileWithDefaultProgram(FullName);
                else
                    FileSystemDataModule.CopyFileToStream(FullName, destination);
            } catch (Exception exc) {
                throw new UserFriendlyException(exc);
            }
        }
        [Persistent]
        public int Size {
            get { return GetPropertyValue<int>("Size"); }
            private set { SetPropertyValue<int>("Size", value); }
        }
        #endregion
        #region IEmptyCheckable Members
        public bool IsEmpty {
            get { return !File.Exists(FullName); }
        }
        #endregion
        #region ISupportFullName Members
        [Size(260), Custom("AllowEdit", "False")]
        public string FullName {
            get { return GetPropertyValue<string>("FullName"); }
            set { SetPropertyValue<string>("FullName", value); }
        }
        #endregion
    }
}